extensions/native/compiler/README.md

```diff
@@ -8,44 +149,59 @@ The naming of `builder.eval` is semantically incorrect because creating a new va
 
-`builder.eval_expr` returns a **[right value](https://www.oreilly.com/library/view/c-in-a/059600298X/ch03s01.html#:~:text=The%20term%20rvalue%20is%20a,are%20close%20to%20the%20truth.)**,
-which is either a constant or a variable. An instruction can use a right value for reading a value directly.
+`builder.eval_expr` returns a **[right value](https://www.oreilly.com/library/view/c-in-a/059600298X/ch03s01.html#:~:text=The%20term%20rvalue%20is%20a,are%20close%20to%20the%20truth.)**, which is either a constant or a variable. `builder.eval_expr` 
+avoids creating a new variable if possible. An instruction can use a right value for reading a value directly.
 
-Currently `builder.eval_expr` only supports `SymbolicVar`.
+Currently `builder.eval_expr` is not generic. Its input must be `SymbolicVar`. `RVar` is the type of the return value of
+`builder.eval_expr`.
 
-## RVar
-`RVar` represents the right value of `SymbolicVar`.
+# AsmCompiler
 
-# CR Variable
-To unify static/dynamic programs in eDSL, we introduce a new concept, CR variable("C" for compile time and "R" for
-runtime). A CR variable behaves like a normal variable in eDSL(e.g. it supports assignment), but it could be either a 
-constant or a variable in runtime.
+ASM compiler is almost a normal compiler except it doesn't use registers(usually address space 3 in RISC-V) and only 
+uses a dedicated address space(4) as memory. It compiles a DSL program into a OpenVM `Program` with a list of native extension instructions.
+
+## Memory Layout
+
+Asm Compiler uses a common memory layout:
+
+|          |
+|----------|
+| STACK    |
+| A0       |
+| HEAP_PTR |
+| STACK    |
+
+`HEAP_PTR`/`A0` are 2 reserved addresses for memory allocation operations. Because heap grows downwards and stack grows 
+upwards, regular operations in DSL cannot touch `HEAP_PTR`/`A0` unless users intend to do so.
+
+Start addresses of each segment are constants in `src/asm/compiler.rs`.
+
+## Stack Variable
+Frame pointer of each basic data type variable can be computed based on their IDs: 
+- `Var`s are stored in stack positions 1+8*0, 2+8*0, 1+8*1, 2+8*1, 1+8*2, 2+8*2, ...
+- `Felt`s are stored in stack positions 3+8*0, 4+8*0, 3+8*1, 4+8*1, 3+8*2, 4+8*2, ...
+- `Ext`s are stored in stack positions 5-8, 13-16, 21-24, ...
 
-Currently, there are 2 types of CR variables: 
+Because DSL doesn't support functions, all pointers of variables are known at compile time. ASM compiler can avoid stack
+overflow at compile time.
 
-## Usize
-`Usize` could be a constant native field or a `Var` at right time.
+## Memory Allocation
+Memory allocation is done by moving the value at `HEAP_PTR` forward with the help of `A0`. Currently, there is no 
+de-allocation at compiler-level. But users can de-allocate by manually resetting the value at `HEAP_PTR`.
 
-## Array
-`Array` has 2 variants: `Fixed` and `Dyn`. 
+The top of the memory is `2^29`. The limit comes from ISA properties. Trying to allocate more than available memory will
```

Comment:
I think `ptr_max_bits` is actually configurable at ISA level. Do we need to enforce that to use AsmCompiler, the VM needs to be configured with `ptr_max_bits=29`?